File: packages/gg-sandbox-ruby/src/gg/tools/tasks.rb
# frozen_string_literal: true

module GG
  # The `tasks` family: the task DAG.
  #
  # Two lowerings live here, both so a model writes ordinary Ruby instead of a tagged union.
  # `description` on a patch is a three-way edit — leave it out to keep the description, pass `nil`
  # to clear it, pass a string to replace it — which the membrane models as a `text-edit` variant
  # and which `UNCHANGED` is the Ruby spelling of. And a status is a Symbol on this side and the
  # membrane's `in-progress` on the other, because WIT identifiers cannot contain an underscore; a
  # model should never see that seam.
  #
  # @api private
  module Tasks
    # The membrane's budget record, as the model-facing one.
    #
    # @param usage [Object] the wire record
    # @return [TaskUsage] the budget a program reads
    def self.usage(usage)
      TaskUsage.new(
        count: Wire.field(usage, "count"),
        max_tasks: Wire.field(usage, "maxTasks")
      )
    end

    # Add a task to the task DAG and return the task budget.
    #
    # @param id [String] The id you choose for it. It is what every other task call takes, and no
    #   two tasks may share one.
    # @param title [String] A short line naming the work.
    # @param description [String, nil] What the work is, at whatever length is useful.
    # @param blocked_by [Array<String>] The ids of the tasks that must be done before this one.
    #   Defaults to none.
    # @return [TaskUsage] how much of the task budget is now used
    # @raise [ToolError] `:conflict` on a duplicate id or on an edge that would close a cycle.
    def self.add_task(id, title, description: nil, blocked_by: [])
      usage(Wire.call("add_task", "tasks", "addTask", [
                        Wire.record(
                          "id" => id,
                          "title" => title,
                          "description" => description,
                          "blockedBy" => Check.strings("add_task", "blocked_by", blocked_by)
                        )
                      ]))
    end

    # Revise a task's title, description and/or status; supply at least one.
    #
    # @param id [String] The task to revise.
    # @param title [String, nil] The title to replace the old one with. Leave it out to keep the
    #   one it has.
    # @param description [String, nil, Unchanged] The description to replace the old one with; `nil`
    #   clears it, and leaving it out (or passing `UNCHANGED`) keeps the one it has.
    # @param status [Symbol, nil] Where the task now stands — one of `TaskStatus`'s symbols. Leave
    #   it out to keep the status it has.
    # @return [nil] nothing; the revision either happened or raised
    # @raise [ToolError] `:not_found` for an unknown id.
    def self.update_task(id, title: nil, description: UNCHANGED, status: nil)
      Wire.call("update_task", "tasks", "updateTask", [
                  id,
                  Wire.record(
                    "title" => title,
                    "description" => Wire.text_edit(description),
                    "status" => Wire.arm(Check.choice("update_task", "status", status,
                                                      TaskStatus::ALL))
                  )
                ])
      nil
    end

    # Replace a task's whole blocker set; no arguments clears every blocker.
    #
    # @param id [String] The task whose blockers to replace.
    # @param blocked_by [Array<String>] The ids of every task that must now be done before it,
    #   splatted. Passing none clears them all.
    # @return [nil] nothing; the replacement either happened or raised
    # @raise [ToolError] `:not_found` for an unknown id, and `:conflict` when an edge would close a
    #   cycle.
    def self.set_blocked_by(id, *blocked_by)
      Wire.call("set_blocked_by", "tasks", "setBlockedBy",
                [id, Check.strings("set_blocked_by", "blocked_by", blocked_by)])
      nil
    end

    # Mark a task done.
    #
    # Tasks it was blocking become actionable once every one of their blockers is done.
    #
    # @param id [String] The task to mark done.
    # @return [nil] nothing; the task either moved or raised
    # @raise [ToolError] `:not_found` for an unknown id.
    def self.complete_task(id)
      Wire.call("complete_task", "tasks", "completeTask", [id])
      nil
    end

    # Remove a task and every blocker edge pointing at it, and return the task budget.
    #
    # @param id [String] The task to remove.
    # @return [TaskUsage] how much of the task budget is still used
    # @raise [ToolError] `:not_found` for an unknown id.
    def self.remove_task(id)
      usage(Wire.call("remove_task", "tasks", "removeTask", [id]))
    end
  end
end

File: packages/gg-sandbox-python/src/gg/tools/tasks.py
"""The `tasks` family: the task DAG.

Two lowerings live here, both so a model writes ordinary Python instead of a tagged union.
`description` on a patch is a three-way edit — leave it out to keep the description, pass `None` to
clear it, pass a string to replace it — which the membrane models as a `text-edit` variant, and which
`UNCHANGED` is the Python spelling of. And a status is a `TaskStatus` member on this side and the
membrane's `in-progress` on the other, because WIT identifiers cannot contain an underscore; a model
should never see that seam.
"""

from __future__ import annotations

from wit_world.imports import tasks as wire
from wit_world.imports import types as wire_types

from ..errors import _call, _strings
from ..types import UNCHANGED, TaskStatus, TaskUsage, Unchanged


def _text_edit(value: str | None | Unchanged) -> wire_types.TextEdit:
    """Lower the `UNCHANGED` / `None` / string sentinel onto the membrane's three-way text edit."""
    if value is UNCHANGED:
        return wire_types.TextEdit_Keep()
    if value is None:
        return wire_types.TextEdit_Clear()
    return wire_types.TextEdit_Set(value)


def _status(status: TaskStatus | None) -> wire.TaskStatus | None:
    """Lower a model-facing status onto the membrane's, whose `in-progress` carries a hyphen."""
    return None if status is None else wire.TaskStatus[status.name]


def _usage(usage: wire.TaskUsage) -> TaskUsage:
    """The membrane's budget record, as the model-facing one."""
    return TaskUsage(count=usage.count, max_tasks=usage.max_tasks)


def add_task(
    id: str,
    title: str,
    *,
    description: str | None = None,
    blocked_by: list[str] | None = None,
) -> TaskUsage:
    """Add a task to the task DAG and return the task budget.

    Args:
        id: The id you choose for it. It is what every other task call takes, and no two tasks may
            share one.
        title: A short line naming the work.
        description: What the work is, at whatever length is useful.
        blocked_by: The ids of the tasks that must be done before this one. Defaults to none.

    Raises:
        ToolError: `conflict` on a duplicate id or on an edge that would close a cycle.
    """
    return _usage(
        _call(
            wire.add_task,
            wire.TaskInput(
                id=id,
                title=title,
                description=description,
                blocked_by=_strings("add_task", "blocked_by", blocked_by),
            ),
        )
    )


def update_task(
    id: str,
    *,
    title: str | None = None,
    description: str | None | Unchanged = UNCHANGED,
    status: TaskStatus | None = None,
) -> None:
    """Revise a task's title, description and/or status; supply at least one.

    Args:
        id: The task to revise.
        title: The title to replace the old one with. Leave it out to keep the one it has.
        description: The description to replace the old one with; `None` clears it, and leaving it
            out keeps the one it has.
        status: Where the task now stands. Leave it out to keep the status it has.

    Raises:
        ToolError: `not-found` for an unknown id.
    """
    _call(
        wire.update_task,
        id,
        wire.TaskPatch(
            title=title,
            description=_text_edit(description),
            status=_status(status),
        ),
    )


def set_blocked_by(id: str, blocked_by: list[str]) -> None:
    """Replace a task's whole blocker set; an empty list clears every blocker.

    Args:
        id: The task whose blockers to replace.
        blocked_by: The ids of every task that must now be done before it. An empty list clears them
            all.

    Raises:
        ToolError: `not-found` for an unknown id, and `conflict` when an edge would close a cycle.
    """
    _call(wire.set_blocked_by, id, _strings("set_blocked_by", "blocked_by", blocked_by))


def complete_task(id: str) -> None:
    """Mark a task done.

    Tasks it was blocking become actionable once every one of their blockers is done.

    Args:
        id: The task to mark done.

    Raises:
        ToolError: `not-found` for an unknown id.
    """
    _call(wire.complete_task, id)


def remove_task(id: str) -> TaskUsage:
    """Remove a task and every blocker edge pointing at it, and return the task budget.

    Args:
        id: The task to remove.

    Raises:
        ToolError: `not-found` for an unknown id.
    """
    return _usage(_call(wire.remove_task, id))
